[Spree build] Wrong number of arguments for Spree::Config.package_factory

HugsDaniel opened the pull request:
#### What ? Why ?

Closes https://github.com/openfoodfoundation/openfoodnetwork/issues/2538

Some other tests were setting `Spree::Config.package_factory` to a custom class called `TestPackageFactory` defined in a spec file (`app_configuration_spec.rb`) and which takes no arguments. This custom class was later wrongfully used by the tests in `coordinator_spec.rb`, making them fail. 

I use rspec `around` hook in `app_configuration_spec.rb` to set the prefs, run the example, and reset them to default `Spree::Stock::Package` afterwards.

mkllnk reviewed (changes requested):
Great work Hugo. Your solution is good. Another approach would be to fix the other spec to not change the factory permanently. It's a good rule for all specs to not leak any state. They have to clean up after themselves. That is easier than changing every other spec they could affect.

About changing the rspec syntax: We always do that in the OFN code. But I think that the Spree project is a special case. We want to use the original Spree as much as possible and apply patches on top where we need them.

For the next Spree upgrade, we will need to apply most of our customisation to the next version of Spree. If you change the syntax here, it will probably create a merge conflict later on. That's why I think we should do as little change as possible.

I'm sorry if a lot of my comments are telling you that you need to change your pull request again or that you did extra work that is not needed. But that's all part of the learning! You do really good work. And learning about all our conventions and how it works together is actually harder than coding. ;-)

Keep up the good work and I'm looking forward to your next pull request.

HugsDaniel commented:
@mkllnk what do you think of the changes ?

mkllnk commented on core/spec/models/spree/app_configuration_spec.rb:
```diff
@@ -28,7 +28,7 @@
 
     it 'uses the set package factory' do
-      prefs.package_factory = TestPackageFactory
+      allow(prefs).to receive(:package_factory) { TestPackageFactory }
```
With this change we actually don't test the `prefs` object any more. You could replace `prefs` with any other object, even a string and tell it: "If I asked you for package_factory, you say TestPackageFactory." And it will do it.

If we don't want to change anything else, we need an `around` block for this special case. It can save the state of the prefs and restore it. Pau did that for the order updater further down:

```ruby
    around do |example|
      default_decorator = prefs.order_updater_decorator
      example.run
      prefs.order_updater_decorator = default_decorator
    end
```

But all this is just a symptom of the wrong test setup. This is a unit test. We all only want to test the unit (class) `Spree::AppConfiguration` in `core/spec/models/spree/app_configuration_spec.rb`. But this spec file contains an unnecessary dependency on the Rails app and is changing the app's state:
```ruby
let (:prefs) { Rails.application.config.spree.preferences }
```

I haven't tested it, but we could get around this problem by just calling `Spree::AppConfiguration.new` there. If that doesn't work, then it's because Spree stores these values in the database and caches them as well. The solution would be change the preference_store or delete the cache. That sounds like a lot of effort and code changes. So if there is not easy, one-line change that works, I would go with the `around` solution I mentioned first.

In the end, Spree is poorly designed here. One advantage of using test driven development with unit tests is that you need to design independent, interchangable modules in order to test them. And that makes your code better as well.

mkllnk reviewed (commented):
Great, we are getting closer to the solution. :-)

HugsDaniel commented:
@mkllnk I used the around hook and it does the job. However it seems that only the specs in `app_configuration_spec.rb` was making the two later tests fail. The setting of `:package_factory` in `packer_spec.rb` comes after those two tests, so it has no influence on them. It's still setting permanently `package_factory` to a custom class, but since it has no influence on the build passing, I figured we could let it like this (in the spirit of "the fewer the changes, the better"). What do you think ?

mkllnk commented:
Good call, don't fix that ain't broken.

I'm glad the around hook is working. I would do it slightly differently though. In the around hook, I would remember the current factory, run the test and then restore the previous factory. The test in the `it` block doesn't need any modifying then. It can still set the factory it wants. That means that the test makes more sense and that the around hook doesn't care about the kind of factory the test wants to use. This is the code I would put there:

```ruby
    class TestPackageFactory; end

    around do |example|
      default_factory = prefs.package_factory
      example.run
      prefs.package_factory = default_factory
    end

    it 'uses the set package factory' do
      prefs.package_factory = TestPackageFactory
      prefs.package_factory.should eq TestPackageFactory
    end
```

HugsDaniel committed:
Implement around hook to avoid changing permanently the factory state

HugsDaniel commented:
I implemented the solution and cleaned previous commits, just kept the interesting one. That should do the trick, what do you think @mkllnk @luisramos0 ?

luisramos0 commented:
I dont see the review icon on this Spree GH project. Maybe some permissions missing for me here.
Anyway, looks good. Approved. Well done Hugo!

mkllnk commented:
@luisramos0 I added the dev team to this repository. You should have write permissions now.

luisramos0 commented:
yes, thanks @mkllnk
